modules/afsocket/afinet-dest-failover.c

```diff
@@ -0,0 +1,429 @@
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "afinet-dest-failover.h"
+#include "messages.h"
+#include "timeutils.h"
+#include "afinet.h"
+
+#include <iv.h>
+
+struct _AFInetDestDriverFailover
+{
+  GList *servers;
+  GList *current_server;
+  GSockAddr *primary_addr;
+
+  guint probe_interval;
+  guint probes_required;
+  guint probes_received;
+  struct iv_timer timer;
+  struct iv_fd fd;
+
+  LogExprNode *owner_expression;
+  FailoverTransportMapper failover_transport_mapper;
+  gpointer on_primary_available_cookie;
+  afinet_dd_failover_on_primary_available on_primary_available_func;
+};
+
+static const int TCP_PROBE_INTERVAL_DEFAULT = 60;
+static const int SUCCESSFUL_PROBES_REQUIRED_DEFAULT = 3;
+
+void
+afinet_dd_failover_set_tcp_probe_interval(AFInetDestDriverFailover *self, gint tcp_probe_interval)
+{
+  self->probe_interval = tcp_probe_interval;
+}
+
+void
+afinet_dd_failover_set_successful_probes_required(AFInetDestDriverFailover *self, gint successful_probes_required)
+{
+  self->probes_required = successful_probes_required;
+}
+
+static GList *
+_primary(AFInetDestDriverFailover *self)
+{
+  return g_list_first(self->servers);
+}
+
+static void
+_afinet_dd_hand_over_connection_to_afsocket(AFInetDestDriverFailover *self)
+{
+  self->probes_received = 0;
+  self->current_server = _primary(self);
+  self->on_primary_available_func(self->on_primary_available_cookie, self->fd.fd, self->primary_addr);
+  self->primary_addr = NULL;
+  self->fd.fd = -1;
+}
+
+static void
+_afinet_dd_start_failback_timer(AFInetDestDriverFailover *self)
+{
+  struct timespec now;
+  glong elapsed_time;
+
+  clock_gettime(CLOCK_MONOTONIC, &now);
+  elapsed_time = timespec_diff_msec(&now, &(self->timer.expires));
+  self->timer.expires = now;
+
+  if (elapsed_time < (self->probe_interval*1000))
+    {
+      timespec_add_msec(&self->timer.expires, (self->probe_interval*1000 - elapsed_time));
+    }
+  iv_timer_register(&self->timer);
+}
+
+static void
+_afinet_dd_tcp_probe_succeded(AFInetDestDriverFailover *self)
+{
+  self->probes_received++;
+  msg_notice("Probing primary server successful",
+             evt_tag_int("successful-probes-received", self->probes_received),
+             evt_tag_int("successful-probes-required", self->probes_required));
+
+  if (self->probes_received >= self->probes_required)
+    {
+      msg_notice("Primary server seems to be stable, reconnecting to primary server");
+      _afinet_dd_hand_over_connection_to_afsocket(self);
+    }
+  else
+    {
+      close(self->fd.fd);
+      g_sockaddr_unref(self->primary_addr);
+      _afinet_dd_start_failback_timer(self);
+    }
+}
+
+static void
+_afinet_dd_tcp_probe_failed(AFInetDestDriverFailover *self)
+{
+  self->probes_received = 0;
+  g_sockaddr_unref(self->primary_addr);
```

Comment:
`self->primary_addr` should be set to NULL to avoid a double-free in `afinet_dd_failover_free()`.